feature to have python output actual alignment

evanbiederstedt opened the pull request:
This is a (flawed) start, related to the following issue: https://github.com/Martinsos/edlib/issues/127 (and this duplicate: https://github.com/Martinsos/edlib/issues/129)

This uses @yech1990 's example to parse the extended cigar (which is just a compressed version of "alignment" from cpp edlib) from the Cython script `edlib.pyx`

> : A method cigarToNiceAlignment that would take cigar returned by align + whatever else it needs (query and target, maybe mode I am not sure at the moment, check cpp implementation of NICE) and returns string which is NICE alignment. That should be it! And it should mostly be code you already wrote here, just slightly modified

I didn't quite do this, but just revised the `align()` method. 

The problem:

As far as I can tell, `ccigar` is always empty. 

I've tried in the Cython code printing these variables:

```
string_alignmentLength = <bytes> cresult.alignmentLength
```

It appears that `cresult.alignmentLength` is also empty. `cedlib.EDLIB_CIGAR_EXTENDED` is some sort of integer? 

These are the issues I'm running into trying to finish this PR

evanbiederstedt commented:
I'm actually not sure what `cresult.alignment` is---I've tried to access it a few ways, and I get segmentation faults. 

Note that I also just created more fields in the result of `align()`. That appears to be a flexible solution.

Martinsos commented:
@evanbiederstedt, it is nice to see that you are doing progress with this :)!

Since python library is working correctly for me, I am guessing the problem might be that you are not running it with correct input parameters: you need to set `task` to "path" if you want edlib to return alignment/cigar. If it is not set to "path", ccigar and cresult.alignment will be empty.
Check the function comments for param `task` and it should be clear.

Regarding the approach itself, while it is not bad, I am not a fan of putting more unnecessary logic into this main function. The way you did it, it is not optional to choose if NICE alignment should be constructed or not.
I would highly prefer having separate function as I mentioned before, so that user can easily use if they want + it does not further complicate / slow down the main function.

evanbiederstedt commented:
> Regarding the approach itself, while it is not bad, I am not a fan of putting more unnecessary logic into this main function. The way you did it, it is not optional to choose if NICE alignment should be constructed or not.
I would highly prefer having separate function as I mentioned before, so that user can easily use if they want + it does not further complicate / slow down the main function.

This is a very good point. It's quite possible that my additions would slow down the method. Yes, I agree with this---I'll implement this as a separate function. 

>  I am not sure what is it that user needs, if you think it is better to separate it upfront, we can go with that. In that case, I would explain that well in the comments and also return structure with nice property names for each of those 3 segments, not just an array/triple.

I think four fields make sense, i.e. four key-value pairs in a dictionary. I think `NICE`, `ref_aln`, `match_aln`, `query_aln` should work----naturally, I'm happy to revise the names.

evanbiederstedt commented:
Hi @Martinsos 

> Since python library is working correctly for me, I am guessing the problem might be that you are not running it with correct input parameters: you need to set task to "path" if you want edlib to return alignment/cigar. If it is not set to "path", ccigar and cresult.alignment will be empty.
Check the function comments for param task and it should be clear.

This is 100% correct. The source code of the current PR  works if `task="path"`. 

e.g. 

```
>>> import edlib
>>> edlib.align(query="AAGGGGTCTCATATC", target="TAAGGATGGTCCCATTC", task="path")
## outputs
{'match_aln': ' ||||  ||||.||| ||', 'ref_aln': 'TAAGGATGGTCCCAT TC', 'query_aln': ' AAGG  GGTCTCATATC', 'editDistance': 5, 'cigar': u'1D4=2D4=1X3=1I2=', 'locations': [(0, 16)], 'alphabetLength': 4, 'NICE': 'TAAGGATGGTCCCAT TC\n ||||  ||||.||| ||\n AAGG  GGTCTCATATC'}
```

So, I'll revise this into a separate function `cigarToNiceAlignment`, and pass it over for review. There may be issues with code duplication, etc. 

Thanks, Evan

evanbiederstedt commented:
Hi @Martinsos 

So, I revised the previous proposal, and implemented `cigarToNiceAlignment()`. 

(I'm not sure `cigarToNiceAlignment()` is the best name, but I cannot think of another. Maybe `niceAlign()`?)

Currently, thinking for potential use cases, I believe I would want the cigar, the NICE alignment, and each component of the NICE alignment. 

I'm tempted to suggest we use `-` instead of spaces to show how the two inputs align. 

That is, instead of


    TAAGGATGGTCCCAT TC
     ||||  ||||.||| ||
     AAGG  GGTCTCATATC


I wonder if users would  prefer

    TAAGGATGGTCCCAT-TC
     ||||  ||||.||| ||
    -AAGG--GGTCTCATATC

As it's a bit more clear where the InDels are located. 

I also changed `target` to `reference`, but I don't mind either term. 

Please revise documentation/variable names as you see fit---we can then discuss updating the README. Thanks!

evanbiederstedt commented:
After thinking about it a bit, I do think that at least `ref_aln`  and `query_aln` should use a different symbol in place of spaces. 

I believe something like:

```
    alignments = {
        'cigar': cigar, 
        'NICE' : "{}\n{}\n{}".format(ref_aln, match_aln, query_aln),  ## f-strings only available from Python 3.6+ sadly
        'ref_aln' : ref_aln.replace(' ', '-'), 
        'match_aln': match_aln,
        'query_aln': query_aln.replace(' ', '-')

    }
```

would result in these strings having `-`. 

Let me know what you think :)

Martinsos commented on bindings/python/edlib.pyx:
```diff
@@ -152,3 +154,98 @@ def align(query, target, mode="NW", task="distance", k=-1, additionalEqualities=
+    cigar = result["cigar"]
+
+    ## cigar parsing, from yech1990
+    for step, code in re.findall("(\d+)(\D)", cigar):
```
step, code -> those names don't seem to be exactly on point. If we have 3X, I would say 3 is number of occurrences (what would step mean?) and X is alignment operation / symbol (code is not so terrible, but still not completely on spot).

evanbiederstedt replied:
I took the above variables directly from @yech1990

How about `occurrence, symbol in re.findall("(\d+)(\D)", cigar):`? Whatever you choose here will be suitable for me

Martinsos replied:
Ah yes I assumed you took them from there, but now they are under scrutiny :D.
I don't care about exact name, as long as it describes what it is.
Let's see: The first thing I would say is "Number of occurrences of the alignment operation". "occurence" is in right direction, but a lot of info is still missing, it does not even sound like a number, and we are dealing with a number here, so I think we can do better. "num_occurences" for example sounds better to me. We could go further with "num_alignment_operation_occurences" but that is really long and since we will describe the second variable with nice name, I think it will be redundant. So "num_occurences" sounds best.
The second thing I would say is "Cigar symbol/code that represent an alignment operation". Since we already know we are in cigar, most information is probably delivered with "alignment_operation", so I would probably go for that.

evanbiederstedt replied:
I like the variable names, yes!

`for num_occurrences, alignment_operation in re.findall("(\d+)(\D)", cigar):
`

And a short comment will clarify any possible ambiguity. This looks good!

Martinsos replied:
Great :)!
As mentioned in "Clean code", comments should be used only when we failed to express what we wanted through the code :D, but I think in this case I agree comment can bring extra clarity and trying to express that through code just does not feel like reasonable effort.

Martinsos commented on bindings/python/edlib.pyx:
```diff
@@ -152,3 +154,98 @@ def align(query, target, mode="NW", task="distance", k=-1, additionalEqualities=
+
+    ## access locations, a list of a set [(start, end)], from yech1990
+    ## ref_pos = locations[0][0]
+    ref_pos = result["locations"][0][0]
```
what is ref_pos -> let's use full words and let's try to be more descriptive.

Martinsos replied:
Aha, ref is for reference. Let's stick to "target"!

Martinsos replied:
And if we are considering how to call those three string, I would suggest:
target_aligned, alignment_operations, query_aligned. Or something in that direction.

evanbiederstedt replied:
> And if we are considering how to call those three string, I would suggest:
target_aligned, alignment_operations, query_aligned. Or something in that direction.

This seems reasonable. I was just using @yech1990 's  variable names. 

I'll use `target` terminology  instead of reference, that's fine.

Martinsos replied:
Ok great!

Martinsos commented on bindings/python/edlib.pyx:
```diff
@@ -152,3 +154,98 @@ def align(query, target, mode="NW", task="distance", k=-1, additionalEqualities=
+            pass
+
+    alignments = {
+        'cigar': cigar, 
```
I would output only those three strings, cigar they already have so no need to return it, and 'NICE' they can construct by joining the three strings if they want.

evanbiederstedt replied:
> I would output only those three strings, cigar they already have so no need to return it, and 'NICE' they can construct by joining the three strings if they want.

Ok, that's reasonable. I'll do this.

evanbiederstedt replied:
Currently only returning these outputs:

```
    alignments = {
        'target_aln' : target_aln, 
        'match_aln': match_aln,
        'query_aln': query_aln
    }
```

Martinsos commented on bindings/python/edlib.pyx:
```diff
@@ -152,3 +154,98 @@ def align(query, target, mode="NW", task="distance", k=-1, additionalEqualities=
+        elif code == "D":
+            ref_aln += reference[ref_pos : ref_pos + step]
+            ref_pos += step
+            query_aln += " " * step
```
You suggested we go with "-" instead of " " -> sounds reasonable, it is easier to read length of gap!
However, one thing to keep in mind is that edlib is not used only by bioinformaticians but also by other people, who sometimes input whole alphabet in it. So question is, which character makes sense? What if "-" is part of the query or target? So we could consider having that as an optional parameter, and default would be "-".

evanbiederstedt replied:
> So we could consider having that as an optional parameter, and default would be "-".

That sounds like a sensible approach. 

Would you recommend some sort of check to make sure users do not use a parameter which isn't suitable? Perhaps "check if character" is too rigid.

evanbiederstedt replied:
Another question:

What would you name this parameter? `gap`?

Martinsos replied:
Hmmm, we could check for them if parameter is not suitable, for example by checking if it is contained in query or target and rejecting it if so. However, that sounds like too much work to me / thinking too much ahead, maybe I would rather just have them pick it for now and leave it to them to figure out what they want to use. If it becomes a problem, we can always add this functionality later.

Naming: `gap` is good! Slightly more informative could be `gapSymbol`, although `gap`is also acceptable for me.

evanbiederstedt replied:
> If it becomes a problem, we can always add this functionality later.

Agreed

> Naming: gap is good! Slightly more informative could be gapSymbol, although gapis also acceptable for me.

I like `gapSymbol` actually

evanbiederstedt commented on bindings/python/edlib.pyx:
```diff
@@ -152,3 +154,98 @@ def align(query, target, mode="NW", task="distance", k=-1, additionalEqualities=
+## function to output NICE alignments
+## only 'task=="path"' works
+
+def cigarToNiceAlignment(query, reference, mode="NW", k=-1, additionalEqualities=None):
```
Hi @Martinsos 

The above looks reasonable. I guess I've been thinking about how I would use this function, though you are correct---developers should have access small, unit functions. 

I'll do the following:

```
query = "elephant"
target = "telephone"
result = align(query, target, task="path")
niceAlignment = cigarToNiceAlignment(result.cigar, query, target, gap = '-')
```

Martinsos replied:
Great :)!
Sorry if I am coming up as somewhat demanding / nitpicking, but we are under no deadline here so I think it is great opportunity to iterate together and try to make this code the best it can be :). And, I have been recently reading "Clean code" book so I am also trying to apply the advice from it.

evanbiederstedt replied:
>  if I am coming up as somewhat demanding / nitpicking

Not at all! No problem whatsoever. :)

I think it makes sense

evanbiederstedt replied:
Hi @Martinsos 
Thanks for the patience on this, and apologies it took forever for me to get back to you. 

I had a slight question:

```
query = "elephant"
target = "telephone"
result = align(query, target, task="path")
niceAlignment = cigarToNiceAlignment(result.cigar, query, target)  # Maybe some more inputs here? Ideally not.
```

It may be that I've forgotten some details, but it looks like we also need the `locations` field from `align()`. My recent commits therefore take the following arguments for `cigarToNiceAlignment()`:

```
def cigarToNiceAlignment(alignResult, query, target, gapSymbol="-"):
```
whereby `alignResult` is the entire output of `align()`. 

Am I missing something? It's quite possible :)

Martinsos commented:
@evanbiederstedt I hope all is fine with you - are you still interested in finalizing this PR? I think we are close to done, would be a shame to let it drop!

evanbiederstedt commented:
Hey @Martinsos 

This is on the to-do list! Apologies so much time has passed---I've been busier than expected. 

Let's see if I can finish this in the next 48 hours. Thanks for checking in though! I am definitely not dropping it. 

Thanks, Evan

Martinsos commented:
Hey no problem, I know how it is! Take your time, just wanted to check if
all is ok :).

On Thu, Jun 27, 2019 at 8:06 PM evanbiederstedt <notifications@github.com>
wrote:

> Hey @Martinsos <https://github.com/Martinsos>
>
> This is on the to-do list! Apologies so much time has passed---I've been
> busier than expected.
>
> Let's see if I can finish this in the next 48 hours. Thanks for checking
> in though! I am definitely not dropping it.
>
> Thanks, Evan
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Martinsos/edlib/pull/132?email_source=notifications&email_token=AALXFBZTCIBQGHZEY2QV2RLP4T6TLA5CNFSM4HN5HJT2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYX5PLA#issuecomment-506451884>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AALXFB7XZEY33IUM73HRXW3P4T6TLANCNFSM4HN5HJTQ>
> .
>

evanbiederstedt commented:
Hi @Martinsos 

Sorry for the delay---it took far longer than 48 hours. 

Please have a look at the recent commits. 

There are a few things to discuss, and then we can finish this up:

(1) I think the function requires both the cigar and `locations` from the output of `align()`, right?

(2) I believe `align()` only outputs a cigar if `task="path"` in `align()`.  If the above outline makes sense, I can write a simple check for this. 

(3) The description of parameters could be improved. 

Let me know how you think we could proceed. Thanks!

evanbiederstedt commented:
Hi @Martinsos 
I think this is close to done. Do you have more recommendations for the finishing touches, so we can push this over the finish line? 

Thanks!

Martinsos commented on bindings/python/edlib.pyx:
```diff
@@ -152,3 +154,91 @@ def align(query, target, mode="NW", task="distance", k=-1, additionalEqualities=
+def cigarToNiceAlignment(alignResult, query, target, gapSymbol="-"):
+    """ Output alignments from align() in NICE format
+    @param {dictionary} alignResult, output of align() 
+            Dictionary with following fields:
```
I would not describe alignResult here again, it is enough that you mentioned it is output of align(), comments of align() can be observed to learn of its shape. We don't want to have to do changes on two places once we do changes to alignResult.

evanbiederstedt replied:
I think that makes sense, removed

Martinsos commented on bindings/python/edlib.pyx:
```diff
@@ -152,3 +154,91 @@ def align(query, target, mode="NW", task="distance", k=-1, additionalEqualities=
+                Here we are using extended cigar format, which uses following symbols:
+                Match: '=', Insertion to target: 'I', Deletion from target: 'D', Mismatch: 'X'.
+                e.g. cigar of "5=1X1=1I" means "5 matches, 1 mismatch, 1 match, 1 insertion (to target)".
+    @param {str or bytes or iterable of hashable objects} query, combined with target must have no more
```
Here, don't describe query like this. Just say it should be query that was used to obtain alignResult. Same goes for target.

evanbiederstedt replied:
Agreed. I also think it needs to be a string, right?

evanbiederstedt replied:
I'll let you check the updates, @Martinsos :)

Martinsos commented on bindings/python/edlib.pyx:
```diff
@@ -152,3 +154,91 @@ def align(query, target, mode="NW", task="distance", k=-1, additionalEqualities=
+            {string} target_aln  target_aln is the aligned target sequence from the input 'target'                
+    """
+
+    ## 'path' - find edit distance, start and end locations and alignment path. (Default)
```
Is this redundant?

evanbiederstedt replied:
Yes, removed

evanbiederstedt replied:
The comments were also outdated

Martinsos commented on bindings/python/edlib.pyx:
```diff
@@ -152,3 +154,91 @@ def align(query, target, mode="NW", task="distance", k=-1, additionalEqualities=
+## function to output NICE alignments
+## only 'task=="path"' works
+
+def cigarToNiceAlignment(alignResult, query, target, gapSymbol="-"):
```
Ok, so the main "news" are that we need position in target to be able to figure out how to correctly construct NICE alignment. This means CIGAR is not enough any more, we also need this position, and it is not so easy to even explain what this position is or where to get it.

So, what you did is, you took the whole alignResult as an argument. I like that, it "hides" what exactly is needed and simplifies things for the user.
However, I would then also change the name of the function, since Cigar is not in the center anymore, and it does not even matter any more that cigar is used as alignment format. So we could rename it to getNiceAlignmentFromAlignResult or alignResultToNiceAlignment or smth like that.

Btw, I suggest you take a look at https://github.com/Martinsos/edlib/blob/master/apps/aligner/aligner.cpp#L331 .
This method does very similar thing, it takes query, target, alignment and position (in CPP but it is similar) and produces NICE alignment. Ok, this method works with alignment while you work with CIGAR, but it is more or less the same thing. Looking at it, it also becomes clear why position is needed. So I would make sure this implementation is aligned (pun intended) with that one.
Btw, one thing I can see from it is that it takes alignment mode (HW, SHW, NW) that was used for alignment. I wonder, do you also need it? Probably you do hm. I think you will want to look into that.

So to summarize: this is great, just rename it to not mention cigar, and probably you will also need to take the alignment mode as parameter -> check that CPP function to see how and why (if needed).

Also, do a check that makes sure CIGAR is set in alignResult, if not throw. And say in comments that alignResult needs to have cigar filled (meaning that task 'path' was used).

evanbiederstedt replied:
Thanks for the feedback @Martinsos 

Would you recommend re-implementing the *cpp code `aligner.cpp`? 

> So to summarize: this is great, just rename it to not mention cigar, and probably you will also need to take the alignment mode as parameter -> check that CPP function to see how and why (if needed).

I've renamed the function. I'm not sure about alignment mode as the parameter....we should put this function in the README

```
import edlib
result = edlib.align(query="TAAGGATGGTCCCATTC", target="AAGGGGTCTCATATC", task="path")
edlib.cigarToNiceAlignment(result, query="TAAGGATGGTCCCATTC", target="AAGGGGTCTCATATC")

{'query_aln': 'TAAGGATGGTCCCAT-TC', 'match_aln': '-||||--||||.|||-||', 'target_aln': '-AAGG--GGTCTCATATC'}
```

> Also, do a check that makes sure CIGAR is set in alignResult, if not throw. And say in comments that alignResult needs to have cigar filled (meaning that task 'path' was used).

This I've tried to do as well. 

Let me know :) Thanks for the help!

Martinsos commented on bindings/python/edlib.pyx:
```diff
@@ -152,3 +154,91 @@ def align(query, target, mode="NW", task="distance", k=-1, additionalEqualities=
+            query_pos += num_occurrences
+            match_aln += gapSymbol * num_occurrences
+        else:
+            pass
```
Can it be something else? This should never happen, right? If so, instead of passing, let's throw. Ohterwise we could hide errors, or produce result for faulty cigar and user will never know.

evanbiederstedt replied:
I don't think it can be anything else, but it's possible there are odd cigar's out there in fields I don't know about. 

I'll throw an error here instead of passing. If there are issues, users should open a github PR.

evanbiederstedt commented on bindings/python/edlib.pyx:
```diff
@@ -152,3 +154,91 @@ def align(query, target, mode="NW", task="distance", k=-1, additionalEqualities=
+           than 256 unique values
+    @param {string} gapSymbol, default "-"
+           String used to represent gaps in the alignment between query and target
+    @return Dictionary with following fields:
```
I tried to provide a better description and example @Martinsos 

Please change and revise as you see fit :) I don't have a great solution here, except to improve the README

evanbiederstedt commented:
Hi @Martinsos 
Please take another look. Feel free to make changes to this branch :)

evanbiederstedt commented on bindings/python/edlib.pyx:
```diff
@@ -152,3 +154,84 @@ def align(query, target, mode="NW", task="distance", k=-1, additionalEqualities=
+    query_pos = 0 ## 0-indexed
+    target_aln = match_aln = query_aln = ""
+
+    cigar = alignResult["cigar"]
```
Good idea. 

I put in two catches: one to check that `alignResult` has an actual CIGAR string, and the other to check that the CIGAR string has contents.

evanbiederstedt replied:
I also put in a check that `alignResult` is the correct data type

evanbiederstedt commented on bindings/python/edlib.pyx:
```diff
@@ -152,3 +154,84 @@ def align(query, target, mode="NW", task="distance", k=-1, additionalEqualities=
+
+
+
+## function to output NICE alignments
```
Very good idea

evanbiederstedt replied:
Accepted

evanbiederstedt commented on bindings/python/edlib.pyx:
```diff
@@ -152,3 +154,84 @@ def align(query, target, mode="NW", task="distance", k=-1, additionalEqualities=
+    @param {string} target, the exact target used for alignResult
+    @param {string} gapSymbol, default "-"
+           String used to represent gaps in the alignment between query and target
+    @return Dictionary with following fields:
```
Accepted

evanbiederstedt commented:
>  However, here in python, you actually use the starting position (right?) which means we don't need to figure out the starting position, and therefore we don't need to know the mode, which is great.

I believe that is right, based on `target_pos = alignResult["locations"][0][0]`, correct? 

I also added a check here such that `locations` exists. 

> The only thing I would ask of you is to confirm that everything works fine for HW by running an example and checking its output.

It looks like it works:

```
>>> import edlib
>>> resultNW = edlib.align(query="TAAGGATGGTCCCATTC", target="AAGGGGTCTCATATC", mode="NW", task="path")
>>> edlib.getNiceAlignmentFromAlignResult(resultNW, query="TAAGGATGGTCCCATTC", target="AAGGGGTCTCATATC")
{'query_aligned': 'TAAGGATGGTCCCAT-TC', 'matched_aligned': '-||||--||||.|||-||', 'target_aligned': '-AAGG--GGTCTCATATC'}
>>> 
>>> resultHW = edlib.align(query="TAAGGATGGTCCCATTC", target="AAGGGGTCTCATATC", mode="HW", task="path")
>>> edlib.getNiceAlignmentFromAlignResult(resultHW, query="TAAGGATGGTCCCATTC", target="AAGGGGTCTCATATC")
{'query_aligned': 'TAAGGATGGTCCCAT-TC', 'matched_aligned': '-||||--||||.|||-||', 'target_aligned': '-AAGG--GGTCTCATATC'}
>>> resultSHW = edlib.align(query="TAAGGATGGTCCCATTC", target="AAGGGGTCTCATATC", mode="SHW", task="path")
>>> 
>>> edlib.getNiceAlignmentFromAlignResult(resultSHW, query="TAAGGATGGTCCCATTC", target="AAGGGGTCTCATATC")
{'query_aligned': 'TAAGGATGGTCCCAT-TC', 'matched_aligned': '-||||--||||.|||-||', 'target_aligned': '-AAGG--GGTCTCATATC'}
```


> I am wondering, does it make sense to add support for the case when query and target are not strings but bytes or iterable hashable objects? We could try to print those also I guess, but that probably does not make much sense. Ok, I am alright with going just with strings for now.

I'm tempted to just use strings for now; if others have requests, we/they can incorporate within this code via PR.


> Ah yes, and one thing is left: while I don't have real tests set up, I do have "poor man's" tests in shape of a script that runs few methods and checks if they failed. So please, add an example or two there, that will test this new function: 

Please add additions to the README; that is very important for the user :)
I think you have a stronger notion of what should be here than me. 


> So please, add an example or two there, that will test this new function: https://github.com/Martinsos/edlib/blob/master/bindings/python/test.py .

Tests are now added :)

evanbiederstedt commented:
Hi @Martinsos 
I also added a few lines to the README

After typing this out, I do think `getNiceAlignmentFromAlignResult()` is too long of method name. Perhaps `getNiceAlignment()` would suffice. I don't feel strongly about this though. 

Please look the changes over, and feel free to make any changes you see fit. 

Thanks, Evan

Martinsos commented on bindings/python/README.rst:
```diff
@@ -63,6 +63,17 @@ Usage
+    result = edlib.align("elephant", "telephone", task="path")  ## users must use 'task="path"' 
+    niceAlign = edlib.getNiceAlignmentFromAlignResult(result, "elephant", "telephone")
+    print(niceAlign['query_aligned'])
+    -elephant
```
```suggestion
```

Martinsos commented on bindings/python/README.rst:
```diff
@@ -63,6 +63,17 @@ Usage
 
+    result = edlib.align("elephant", "telephone", task="path")  ## users must use 'task="path"' 
+    niceAlign = edlib.getNiceAlignmentFromAlignResult(result, "elephant", "telephone")
+    print(niceAlign['query_aligned'])
```
```suggestion
    print(niceAlign['query_aligned'])  # "-elephant"
```

Martinsos commented on bindings/python/README.rst:
```diff
@@ -63,6 +63,17 @@ Usage
+    niceAlign = edlib.getNiceAlignmentFromAlignResult(result, "elephant", "telephone")
+    print(niceAlign['query_aligned'])
+    -elephant
+    print(niceAlign['target_aligned'])
```
```suggestion
    print(niceAlign['matched_aligned'])  # "-|||||.|."
```

Martinsos commented on bindings/python/README.rst:
```diff
@@ -63,6 +63,17 @@ Usage
+    print(niceAlign['query_aligned'])
+    -elephant
+    print(niceAlign['target_aligned'])
+    telephone
```
```suggestion
```

Martinsos commented on bindings/python/README.rst:
```diff
@@ -63,6 +63,17 @@ Usage
+    -elephant
+    print(niceAlign['target_aligned'])
+    telephone
+    print(niceAlign['matched_aligned'])
```
```suggestion
    print(niceAlign['target_aligned'])  # "telephone"
```

Martinsos commented on bindings/python/README.rst:
```diff
@@ -63,6 +63,17 @@ Usage
+    print(niceAlign['target_aligned'])
+    telephone
+    print(niceAlign['matched_aligned'])
+    -|||||.|.
```
```suggestion
```

evanbiederstedt commented:
Hi @Martinsos 

> However, please check the suggestions I made and consider approving them.

I've accepted them all! 

> Also, in README, line 39, it says "Edlib has only one function" -> that is not true anymore. It would be great if you could modify that so it says that there is one main function, and then there is utility function -> the one which you implemented.

Done!

> Before merging, please squash all the commits you did into one commit.

Ok, I could do this.

evanbiederstedt commented:
Hi @Martinsos 

Thanks for the help on this. Everything looks good from my end. I think you need to merge now, as I don't have permissions.

Martinsos commented:
@evanbiederstedt awesome!!
You are right, I forgot I have to merge -> done.
Thanks for all the effort, and I am looking forward to future PRs :D!

evanbiederstedt commented:
@Martinsos  I'm very happy to collaborate again!  (Fork deleted.)
